jme3-core/src/main/java/com/jme3/anim/tween/action/Action.java

```diff
@@ -1,22 +1,112 @@
+/*
+ * Copyright (c) 2009-2021 jMonkeyEngine
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ * * Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ *
+ * * Neither the name of 'jMonkeyEngine' nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
 package com.jme3.anim.tween.action;
 
 import com.jme3.anim.AnimationMask;
+import com.jme3.anim.Armature;
 import com.jme3.anim.tween.Tween;
 import com.jme3.util.clone.Cloner;
 import com.jme3.util.clone.JmeCloneable;
 
+/**
+ * Base implementation (1st level impl/prototype) of the interface #{@link Tween} for the new animation system.
```

Comment:
Again, Nehon has built his animation api based on Paul's Tweens from lemur.

In `Tweens` API, you will notice this pattern : 
Tween (interface)
|
Sequence, Parallel, ....., AbstractTween (Default implementation of tween, resembles functional code aka interpolates transforms)
|
Default Tweens

WHILE in the new system, Nehon build the system on 2 phases of Abstraction levels or layers : 
Tween
|
Action (Non-functioning tween, its a tween-collective phase)
|
BlendableAction (functioning tween), BaseAction (functioning tween, used to gather tweens and interpolate them)
|
ClipAction (the main functioning class), BlendAction (for blending between 2 actions using a blendSpace).

I can prepare and supply a UML diagram for the animation system to understand its core, we can have another discussion on forums for this anyway.

Reply:
I think the production level would be the ClipAction, BaseAction, and BlendAction, those are the ones we use in our code tho.

Reply:
That whole "(impl/prototype)" parenthetical is at best redundant and at worst confusing because "prototype" is an overloaded word that on the whole of language more commonly refers to something different than meant here.  The "Base implementation" and "what it does" are all that's needed.  The extra language doesn't add anything.

(Frankly, if JME wanted to 'adapt' tweens into something with more action-y juice when action should have been an interface with a base implementation.)

Also, you use other software pattern words in an odd way.  "doInterpolate()" is not a factory method by any stretch of the meaning of "factory pattern" that I know of.  A factory creates something.  I suspect you mean a different word here.

To me, Action seems to be an adapter class.  It adapts (one or more) more general Tweens into JME's "Action" which provides more animation-specific settings beyond what a generic tween would provide.  But since it is not implementing a new interface, I guess we could call it a decorator if we wanted to get fancy with our software pattern names.

Reply:
I agree with the name "adapter", but the system uses the abstract factory pattern starting from `BlendableAction` and `doInterpolate()` is a real factory method, you will override to create your interpolation.

Reply:
re: "//override this factory method to do interpolation"

Is it doing an interpolation or is it creating an object?

I actually really don't like the example provided as abstract AbstractAction extends Action (which is already abstract) feels very strange.  And it is encouraging the very very worst kind of class hierarchy (micro-inheritance over simple composition).

Reply:
Its doing the interpolation, since no instance return.

Alright, no problem, I will remove the example code from all classes docs.

Reply:
What about a `Creational pattern` instead of `Abstract factory pattern` ?

Reply:
Factory, creational... they are all the same.  They imply that you are CREATING something.  Which you say you are not.

Factory pattern:
```
public class MyCoolFactory implements SomeFactory {
    // factory method
    public TheThing createTheThing() {
        return new SpecialTheThing();
    }
}
```

If you aren't creating something then it is definitely 100% not a factory pattern.  And "creational" is just another name for "factory" here... and even has "Create" right in the name.

If all you are doing is overriding some behavior then you are overriding some behavior.  There is no special name for that other than "object oriented programming" or "specialization".  You are just extending the API of the class to allow more customization of behavior on some thing partially implemented.  That the specific example is a little upside down doesn't really change what it is.

Reply:
You are right, but I am not sure if that's the case here as `doInterpolate` returns a boolean, if it its false then no interpolation is created, if its true then the interpolation is created.

Reply:
I think just remove the example.  The engine has other examples that extend this class already, no?

Reply:
I will remove the abstract factory pattern keyword from the docs.

Reply:
So, to recap : 
1) Rephrase this class doc including removal of redundant words.
2) Removal of example code.
3) Removal of Abstract factory pattern keyword.

Reply:
@pspeed42 yes there are examples, do you want to document them by a see tag ?